### ISSUE TITLE
Control audio playback from buttons on the home widget

### ISSUE THREAD

**Issue by sanidhya711** [ABausG/home_widget#127]
Hi, I am not able to control audio playback(play/pause/seek_next/seek_prev) from the buttons on the home_widget.

not being able to control playback from buttons on Home_widget means that the callback that i am registering by using 
`  HomeWidget.registerBackgroundCallback(homeScreenWidgetCallback);`
has these lines of code
```
dynamic homeScreenWidgetCallback(Uri? uri) async {
  print(uri?.host);
  if(uri?.host == 'seek_prev'){
    if((player.position.inMilliseconds ?? 0) >= 2100 || player.currentIndex == 0){
      player.seek(Duration(seconds: 0));
    }else{
      player.seekToPrevious();
      player.play();
    }
  }else if (uri?.host == 'play_pause') {
     print("waddup mf");
  if(!player.playing){
    player.play();
  }else{
    player.pause();
  }
  }else{
    if(player.nextIndex  != null){
      player.seekToNext();
      player.play();
    }else{
      player.pause();
      player.seek(Duration.zero, index: player.currentIndex);
    }
  }

  await HomeWidget.updateWidget(name: 'AppWidgetProvider', iOSName: 'AppWidgetProvider');
}
```

but the player method Im using don't actually stop the music or change its state though they do one thing on click that is:
**if some other application is playing audio and I press any of these audio control buttons on my home widget the music stops from the other app.**

I think I might be missing some service element but I tried a lot of stuff and nothing really worked out. Any help is greatly appreciated

**Comment by ABausG** [comment]
Couple of questions:
- I assume this is on Android, as the package does not yet support the interactive Widgets added with iOS 17
- Is the callback invoked at all?
- Can you share the code that shows where you register the callback for the homeWidget?
- Can you share some widget code to show how you are invoking the background callback?
- How is the `player` created/accessed in the code?

**Comment by sanidhya711** [comment]
Hi, thanks for looking into the issue
here are the answers to your questions

- I assume this is on Android, as the package does not yet support the interactive Widgets added with iOS 17

  Yup this is on android, Ive tried running it on two physical devices
    > Samsung Galaxy M12(Android 11)
    > One Plus 7(11)

- Is the callback invoked at all?

  It is called indeed as i can see the logs being printed into the console when i click the button on the home widget

- Can you share the code that shows where you register the callback for the homeWidget?
 
  This is from my main.dart file
```
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  HomeWidget.registerBackgroundCallback(homeScreenWidgetCallback);

  await JustAudioBackground.init(
    notificationColor: Colors.blue,
    androidNotificationChannelId: 'com.ryanheise.bg_demo.channel.audio',
    androidNotificationChannelName: 'Audio playback',
    androidNotificationOngoing: true,
  );

  runApp(const Localize());
}

dynamic homeScreenWidgetCallback(Uri? uri) async {
  print(uri?.host);
  if(uri?.host == 'seek_prev'){
    if((player.position.inMilliseconds ?? 0) >= 2100 || player.currentIndex == 0){
      player.seek(Duration(seconds: 0));
    }else{
      player.seekToPrevious();
      player.play();
    }
  }else if (uri?.host == 'play_pause') {
    if(!player.playing){
      player.play();
    }else{
      player.pause();
    }
  }else{
    if(player.nextIndex  != null){
      player.seekToNext();
      player.play();
    }else{
      player.pause();
      player.seek(Duration.zero, index: player.currentIndex);
    }
  }

  await HomeWidget.updateWidget(name: 'AppWidgetProvider', iOSName: 'AppWidgetProvider');
}

```

- Can you share some widget code to show how you are invoking the background callback?

  This is from my WidgetProvider.kt file
 ```
class AppWidgetProvider : HomeWidgetProvider() {
    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray, widgetData: SharedPreferences) {
        appWidgetIds.forEach { widgetId ->
            val views = RemoteViews(context.packageName, R.layout.widget_layout).apply {

                val pendingIntent = HomeWidgetLaunchIntent.getActivity(context, com.ryanheise.audioservice.AudioServiceActivity::class.java)
                setOnClickPendingIntent(R.id.widget_root, pendingIntent)

                val title = widgetData.getString("title", "Title...")
                val artists = widgetData.getString("artists", "Artists...");
                val paused = widgetData.getBoolean("paused", true);
                val cover = widgetData.getString("cover","");

                val decodedString: ByteArray = Base64.decode(cover, Base64.DEFAULT)
                val decodedByte: Bitmap = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.size)

                setTextViewText(R.id.title, title)
                setTextViewText(R.id.artists, artists)

                if(cover == ""){
                    setImageViewResource(R.id.cover, R.drawable.logo);
                }else{
                    setImageViewBitmap(R.id.cover, decodedByte);
                }

                if(paused){
                    setImageViewResource(R.id.play_pause, R.drawable.play);
                }else{
                    setImageViewResource(R.id.play_pause, R.drawable.pause);
                }

                val seekPrev = HomeWidgetBackgroundIntent.getBroadcast(context, Uri.parse("myAppWidget://seek_prev"))
                setOnClickPendingIntent(R.id.seek_prev, seekPrev)

                val play_pause = HomeWidgetBackgroundIntent.getBroadcast(context, Uri.parse("myAppWidget://play_pause"))
                setOnClickPendingIntent(R.id.play_pause, play_pause)

                val seekNext = HomeWidgetBackgroundIntent.getBroadcast(context, Uri.parse("myAppWidget://seek_next"))
                setOnClickPendingIntent(R.id.seek_next, seekNext)
            }

            appWidgetManager.updateAppWidget(widgetId, views)
        }
    }
}

```

- How is the player created/accessed in the code?

  the player is an instance of  **just_audio: ^0.9.31** and im also using  **just_audio_background: ^0.0.1-beta.9** in my project.

   I am creating the player using these lines of code `final player = AudioPlayer();` just above the main method and these lines are placed outside any method or function so the variable "player" is global


Thank you so much for looking into the issue, let me know if you need any more information or code.

**Comment by ABausG** [comment]
Given that the callback is invoked it seems to me like the code from this package seems to work as expected.

The way the background callbacks work in home_widget that the dart method is called using a new FlutterEngine which means that your `player` variable will be initialised again. 

Not sure if the `just_audio` / `just_audio_background` package offer a way of getting access to a running audio player.

Maybe also try if it works if you call your code to initials `just_audio_background` in the beginning of the `homeScreenWidgetCallback` like this:

```dart
dynamic homeScreenWidgetCallback(Uri? uri) async {
  print(uri?.host);
  await JustAudioBackground.init(
    notificationColor: Colors.blue,
    androidNotificationChannelId: 'com.ryanheise.bg_demo.channel.audio',
    androidNotificationChannelName: 'Audio playback',
    androidNotificationOngoing: true,
  );
  ...
```

---

Another possible workaround might be to send raw playback intents through something like https://pub.dev/packages/android_intent_plus and https://developer.android.com/reference/androidx/mediarouter/media/MediaControlIntent#ACTION_PAUSE()

---

I will close this issue as the problem is not itself from this package but feel free to ask any further questions

**Comment by zhushenwudi** [comment]
I think this is a design flaw. If the process is isolated every time this function is entered, designing the interaction capability of this widget will become meaningless and unable to control the behavior of the main program. I am currently implementing a player implemented by just_audio, and currently all variables of the main process cannot be accessed. @ABausG

**Comment by ABausG** [comment]
This is underlying though how HomeScreenWidgets work as they are not part of the main app but running more in parallel. You are still able to communicate with the isolates e.g through HomeWidgets data layer or other persistent storage solutions to pass around data.

**Comment by zhushenwudi** [comment]
> This is underlying though how HomeScreenWidgets work as they are not part of the main app but running more in parallel. You are still able to communicate with the isolates e.g through HomeWidgets data layer or other persistent storage solutions to pass around data.

Now I try to send a flutter channel message, and achieved the entire pathway
At present, the project does not support apps passing Flutter container objects to the native layer in the background. Perhaps it should be passed to the runner module and then passed to the widget for implementation
 I am actively trying to implement this part of the logic, and I understand that it should be achievable. Perhaps the project just did not implement this part
